quinn-proto/src/connection.rs

```diff
@@ -744,13 +746,13 @@ impl Connection {
     ) -> Option<BytesMut> {
         let mut new_crypto = None;
         let crypto = if partial_decode.is_handshake() {
-            &self.handshake_crypto
+            self.find_crypto(CryptoLevel::Initial, 0).unwrap()
         } else if partial_decode.key_phase() != self.key_phase {
-            new_crypto = Some(self.crypto.as_ref().unwrap().update());
+            new_crypto = Some(self.cryptos.back().unwrap().2.update());
             new_crypto.as_ref().unwrap()
         } else {
-            if let Some(ref crypto) = self.crypto {
-                crypto
+            if let Some(crypto) = self.cryptos.back() {
+                &crypto.2
             } else {
```

Comment:
`self.cryptos` is never empty, so this branch will never be executed.

Reply:
Hmm, so here's what I'm wondering about: we create new packet number encryption keys for key phase changes, but if packets arrive out of order (which we can discern only based on the packet number), we will use the incorrect keys to decrypt the packet number. Does that mean packet number protection keys don't get updated alongside the other keys? Or am I missing some other solution to this problem?

Reply:
That's been on my mind too since reviewing this. The draft doesn't seem to mention the problem, but my reading is that the PNE key does get updated, and quant and ngtcp2 (for example) both seem to update it. One solution might be: if a packet with a mismatched key phase fails authentication with key set n+1, retry the whole procedure including packet number decryption using key set n-1.

Reply:
Word on slack is that PNE keys will not be affected by key update in draft 17.